### PR TITLE
Sdk 3.x.x patch branch 

### PR DESCRIPTION
**Pull request by Jeasmine**
This Branch/PR is based on the last 3.X.X version before the 4.0.0 Major Release was merged. I have cherry-picked all non-merge commits except Carousel Page Impressions to this version. This branch/PR can be used to maintain a 2.X.X version to be used in wrapper SDKs

## 
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1269)

### CODE REVIEW

**Review by jkasten2 — APPROVED** [review]

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/ActivityLifecycleHandler.java [hunk 1]
```diff
@@ -42,95 +43,53 @@
...
+    void onActivityCreated(Activity activity) {
     }
 
-    static void onActivityStarted(Activity activity) {
+    void onActivityStarted(Activity activity) {
     }
 
-    static void onActivityResumed(Activity activity) {
+    void onActivityResumed(Activity activity) {
+        OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "onActivityResumed: " + activity);
```
Comment (aimiebell) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/ActivityLifecycleHandler.java:85]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSOutcomeEventsController.java [hunk 2]
```diff
@@ -204,13 +204,17 @@ private void sendAndCreateOutcomeEvent(@NonNull final String name,
                     break;
                 case DISABLED:
                     OneSignal.Log(OneSignal.LOG_LEVEL.VERBOSE, "Outcomes disabled for channel: " + influence.getInfluenceChannel());
```
Comment (aimiebell) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSOutcomeEventsController.java:206]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java [hunk 3]
```diff
@@ -1252,6 +1271,7 @@ static void logHttpError(String errorString, int statusCode, Throwable throwable
    @WorkerThread
    static void onAppLostFocus() {
       Log(LOG_LEVEL.DEBUG, "Application lost focus");
```
Comment (aimiebell) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java:1273]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java [hunk 4]
```diff
@@ -251,40 +261,49 @@ private void calculateHeightAndShowWebViewAfterNewActivity() {
 
         OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "In app message new activity, calculate height and show ");
```
Comment (aimiebell) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java:262]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java [hunk 5]
```diff
@@ -171,6 +171,46 @@ public void testDisableInAppMessagingPreventsMessageDisplay() throws Exception {
...
+    @Test
+    public void testMessagesDoNotDisplayPastEndTime() throws Exception {
+        final OSTestInAppMessage message = InAppMessagingHelpers.buildTestMessageWithEndTime(OSTriggerKind.CUSTOM, "test_key", OSTestTrigger.OSTriggerOperator.EQUAL_TO.toString(), 3, true);
+        setMockRegistrationResponseWithMessages(new ArrayList<OSTestInAppMessage>() {{
+            add(message);
+        }});
+
+        // the SDK should read the message from registration JSON, set up a timer, and once
+        // the timer fires the message should get shown.
+        OneSignalInit();
```
Comment (aimiebell) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/InAppMessageIntegrationTests.java:183]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/MainOneSignalClassRunner.java [hunk 6]
```diff
@@ -1748,6 +1749,27 @@ public void shouldLogoutOfEmail() throws Exception {
       assertEquals("b2f7f966-d8cc-11e4-bed1-df8f05be55ba", logoutEmailPost.payload.get("app_id"));
    }
 
+   @Test
+   public void logoutEmailShouldNotSendEmailPlayersRequest() throws Exception {
+      // 1. Init OneSignal and set email
+      OneSignalInit();
```
Comment (aimiebell) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/MainOneSignalClassRunner.java:1755]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

**Pull request closed without merging**